test/e2e/test/elasticsearch/checks_budget.go

```diff
@@ -116,27 +116,39 @@ func (c *ChangeBudgetCheck) Stop() {
 	c.stopChan <- struct{}{}
 }
 
-func (c *ChangeBudgetCheck) Verify(esSpec v1beta1.ElasticsearchSpec) error {
-	desired := esSpec.NodeCount()
-	budget := esSpec.UpdateStrategy.ChangeBudget
-
+func (c *ChangeBudgetCheck) Verify(from v1beta1.ElasticsearchSpec, to v1beta1.ElasticsearchSpec) error {
+	desired := to.NodeCount()
+	budget := to.UpdateStrategy.ChangeBudget
+
+	// allowedMin, allowedMax bound observed values between the ones we expect to see given desired count and change budget.
+	// seenMin, seenMax allow for ramping up/down nodes when moving from spec outside of <allowedMin, allowedMax> node count.
+	// It's done by tracking lowest/highest values seen outside of bounds. This permits the values to only move monotonically
+	// until they are inside <allowedMin, allowedMax>.
```

Comment:
Not sure I understand the last comment. With `allowedMax=2` and `seenMin=5` initially would 5,3,5,3,2 not be a valid sequence of observations that is not monotonic?

Reply:
Ah, I thought it's not a valid sequence :) if we allow max 2 and we are at 5 we should only allow decreasing replicas, not creating new ones. It should be blocked at:
https://github.com/elastic/cloud-on-k8s/blob/bb501e9a0ce3359736078ac3cf15362f3122c4c0/pkg/controller/elasticsearch/driver/upscale_state.go#L176

So in this case if we were at [5,>3<,5,3,2] we should either stay at 3 or move to 2.

Reply:
You are right, all good!